pallets/phala/src/stakepool.rs

```diff
@@ -540,6 +562,110 @@ pub mod pallet {
 
 			Ok(())
 		}
+		/// add a staker accountid to contribution whitelist.
+		/// calling this method will forbide stakers contribute who isn't in the whitelist.
+		/// caller must be the owner of the pool.
+		/// if a pool hasn't registed in the wihtelist map, any staker could contribute as what they use to do.
+		/// whitelist has a lmit len of 500 stakers.
+		#[pallet::weight(0)]
+		pub fn add_staker_to_whitelist(
+			origin: OriginFor<T>,
+			pid: u64,
+			staker: T::AccountId,
+		) -> DispatchResult {
+			let owner = ensure_signed(origin)?;
+			let mut pool_info = Self::ensure_pool(pid)?;
+			ensure!(pool_info.owner == owner, Error::<T>::UnauthorizedPoolOwner);
+			if let Some(mut whitelist) = LimitContribtionPoolsList::<T>::get(&pid) {
+				if !whitelist.contains(&staker) {
+					ensure!(whitelist.len() as u32 <= MAX_WHITELIST_LEN, Error::<T>::ExceedWhitelistMaxLen);
+					whitelist.push(staker.clone());
+					LimitContribtionPoolsList::<T>::insert(&pid, &whitelist);
+					Self::deposit_event(Event::<T>::PoolWhitelistStakerAdded{
+						pid,
+						staker,
+					});
+				}
+			} else {
+				let mut new_list = vec![staker.clone()];
+				LimitContribtionPoolsList::<T>::insert(&pid, &new_list);
+				Self::deposit_event(Event::<T>::PoolWhitelistCreated {
+					pid,
+				});
+				Self::deposit_event(Event::<T>::PoolWhitelistStakerAdded{
+					pid,
+					staker,
+				});
+			}
+			Ok(())
+		}
+		/// remove a staker accountid to contribution whitelist.
+		/// caller must be the owner of the pool.
+		/// if the last staker in the whitelist is removed, the pool will return back to a normal pool that allow anyone to contribute.
+		#[pallet::weight(0)]
+		pub fn remove_staker_from_whitelist(
+			origin: OriginFor<T>,
+			pid: u64,
+			staker: T::AccountId,
+		) -> DispatchResult {
+			let owner = ensure_signed(origin)?;
+			let mut pool_info = Self::ensure_pool(pid)?;
+			ensure!(pool_info.owner == owner, Error::<T>::UnauthorizedPoolOwner);
+			if let Some(mut whitelist) = LimitContribtionPoolsList::<T>::get(&pid) {
+				if whitelist.contains(&staker) {
+					&whitelist.retain(|accountid| accountid != &staker);
+					if whitelist.len() as u32 <= 0 {
+						LimitContribtionPoolsList::<T>::remove(&pid);
+						Self::deposit_event(Event::<T>::PoolWhitelistStakerRemoved {
+							pid,
+							staker: staker.clone(),
+						});
+						Self::deposit_event(Event::<T>::PoolWhitelistDeleted {
+							pid,
+						});
+					} else {
+						LimitContribtionPoolsList::<T>::insert(&pid, &whitelist);
+						Self::deposit_event(Event::<T>::PoolWhitelistStakerRemoved {
+							pid,
+							staker: staker.clone(),
+						});
```

Comment:
This will be executed anyway. Why not put it outside the if statement?

Reply:
done